alloc: Use C11 aligned_alloc

jdemel opened the pull request:
We require C11. Thus, we can now rely on `aligned_alloc` and `free`.
[aligned_alloc](https://en.cppreference.com/w/c/memory/aligned_alloc)
Also, the documentation is updated accordingly.

Closes #193

jdemel committed:
alloc: Use C11 aligned_alloc

We require C11. Thus, we can now rely on `aligned_alloc` and `free`.
https://en.cppreference.com/w/c/memory/aligned_alloc

michaelld commented on lib/volk_malloc.c:
```diff
@@ -29,102 +29,30 @@
 /*
  * For #defines used to determine support for allocation functions,
  * see: http://linux.die.net/man/3/aligned_alloc
+ *
```
I think the just above comment and link can now be removed too.

michaelld commented on lib/volk_malloc.c:
```diff
@@ -29,102 +29,30 @@
-  // "alignment must be a power of two and a multiple of sizeof(void *)"
-  // volk_get_alignment() could return 1 for some machines (e.g. generic_orc)
-  if (alignment == 1)
+  if (alignment == 1){
```
(1) Thinking to just remove this statement, since one can use `aligned_alloc` with alignment of 1. I'd guess internal to that function it checks the alignment & if == 1 then uses `malloc`, so I'd bet we're not saving a lot of CPU cycles here. And, it cleans up our code.
(2) Thinking making both `volk_malloc` and `volk_free` inline functions, since they are so short to begin with; no real need to have them be proper functions, IMHO. Maybe `__VOLK_ATTR_INLINE` since that already exists in `volk_common.h`? If not that, then we can define a `__VOLK_FUNC_INLINE` or something like it & use that here.

jdemel replied:
Since GCC throws this warning, I'd suggest we don't inline this function. 
```
warning: always_inline function might not be inlinable [-Wattributes]
 __VOLK_ATTR_INLINE void volk_free(void *ptr)
                         ^~~~~~~~~
warning: always_inline function might not be inlinable [-Wattributes]
 __VOLK_ATTR_INLINE void *volk_malloc(size_t size, size_t alignment)
```

That would also be consistent with its current implementation. Further, I'd assume current compilers inline such functions whenever appropriate and possible.

michaelld commented on lib/volk_malloc.c:
```diff
@@ -29,102 +29,30 @@
 
-#endif // _POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600 || HAVE_POSIX_MEMALIGN
 
 //#endif // _ISOC11_SOURCE
```
I'd say to remove these as well. They seem like cruft with all of the macros above removed.

michaelld reviewed (changes requested):
Good first shot! A few changes I think are in order. Or at least discussion of how we might remove more code and other comments.

rear1019 commented on include/volk/volk_malloc.h:
```diff
@@ -32,22 +32,8 @@ __VOLK_DECL_BEGIN
- * Because of the ways in which volk_malloc may allocate memory, it is
- * important to always free volk_malloc pointers using volk_free.
+ * Internally, we use C11 `aligned_alloc` to allocate aligned memory.
+ * Thus, we rely on C11 syntax and compilers.
```
MSVC never fails to disappoint when it comes to (lacking) support of “old” features [1]:
> C11 The Universal CRT implemented the parts of the C11 Standard Library that are required by C++17, with the exception of C99 strftime() E/O alternative conversion specifiers, C11 fopen() exclusive mode, and C11 aligned_alloc(). The latter is unlikely to be implemented, because C11 specified aligned_alloc() in a way that's incompatible with the Microsoft implementation of free(): namely, that free() must be able to handle highly aligned allocations.

In other words: Special handling for MSVC is required.

[1] https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019

jdemel replied:
It is really disappointing. Thanks for the link.
I added a workaround, i.e. a preprocessor macro. I really don't like it but as long as we want non-standard compliant compilers to be able to build VOLK that's what we have to do.

jdemel committed:
alloc: Fix for incomplete MSVC standard compliance

MSVC is not standard compliant and probably will never be.
In case `_MSC_VER` is detected, instead of C11 `aligned_alloc` and `free`,
we must use `_aligned_malloc` and `_aligned_free`.

jdemel commented:
Thanks for your comments. I updated this PR accordingly. Apparently, CI passes again.

carlesfernandez commented:
I think function `print_malloc()` in [volk_config_info.cc](https://github.com/gnuradio/volk/blob/d3e9614cb9b0f5b2f65a7f9b77ede7421557c530/apps/volk-config-info.cc#L39) should be changed to something like:

```
void print_malloc()
{
    // You don't want to change the volk_malloc code, so just copy the if/else
    // structure from there and give an explanation for the implementations
    std::cout << "Used malloc implementation: ";
#if defined(_MSC_VER)
    std::cout << "aligned_malloc" << std::endl;
#else
    std::cout << "aligned_alloc" << std::endl;
#endif
}
```

carlesfernandez commented:
Also, [lines 68 to 78](https://github.com/gnuradio/volk/blob/d3e9614cb9b0f5b2f65a7f9b77ede7421557c530/lib/CMakeLists.txt#L68-L78) in `lib/CMakelists.txt` should be removed, since `HAVE_POSIX_MEMALIGN` is not used elsewhere.

michaelld commented:
Good catches, @carlesfernandez ! Yes, let's get those changes into this PR as well. They all fit together nicely.

jdemel committed:
alloc: update to reflect alloc changes

This commit rewrites `print_malloc` to print the newly available methods.
Remove obsolete checks for `posix_malloc`.

jdemel commented:
Thanks for those hints @carlesfernandez . I pushed an update with your proposed changes.

carlesfernandez commented:
I'm getting segfaults when trying this on macOS.  Apparently, aligned_alloc is not available in Apple's C11  :-(

carlesfernandez commented:
Please check https://github.com/carlesfernandez/volk/tree/aligned-alloc-macos for a solution that works for macOS

michaelld commented:
Well carp. Let me test on my macOS box (10.14) ...

michaelld commented:
Indeed there is no `aligned_alloc` on macOS 10.14 latest Xcode, which means it just doesn't exist on macOS. Hence: instead of assuming it works, let's add a test for it & augment the current `volk_malloc` codes to use it if found, but otherwise do whatever it does right now. I think this means we'll need to check in `volk_free` for if the argument is NULL or not & deal with it accordingly.

jdemel commented:
@carlesfernandez thanks for your hints.

[Our old doc reference](https://linux.die.net/man/3/aligned_alloc) states, that `posix_memalign`'ed pointers can just be passed to `free`.
Thus, I'd really want to keep the `volk_free` function as simple as possible. Only MSVC gets a special case.

I looked through the proposed changes. (sidenote: I'm all in favor of auto-formatting code but it makes it harder to get to the essentials of the commit.) 

GCC or Clang will return True for this call:
```
include(CheckSymbolExists)
check_symbol_exists(posix_memalign stdlib.h HAVE_POSIX_MEMALIGN)
```
But usually, we want to use `aligned_alloc` which is also defined in [`stdlib.h`](https://en.cppreference.com/w/c/memory/aligned_alloc).
Thus, I'll check for a solution that prefers `aligned_alloc` and just uses `posix_memalign` as a fallback.

jdemel committed:
alloc: Implement fall backs for C11 aligned_alloc

Apple Clang and MSVC do not support C11 aligned_alloc. Thus, we need
fall backs. With this commit, we should have all fall backs in place.

jdemel commented:
So, all tests pass with the newest version. Unfortunately, we don't have CI for macos yet. Thus, I'd like to ask you to test the newest version, @carlesfernandez @michaelld .
I tried to follow @carlesfernandez solution but with some differences that reflect my considerations.

carlesfernandez commented:
Hi @jdemel, I confirm that your solution works on macOS 10.15, latest Xcode.

michaelld reviewed (approved):
LGTM

michaelld commented:
@rear1019 do these changes work for you / on Windows?

carlesfernandez commented:
Damn! 

In macOS, ```CHECK_SYMBOL_EXISTS(aligned_alloc stdlib.h USE_ALIGNED_ALLOC)``` sets  ```USE_ALIGNED_ALLOC``` to ```1```,  so it happily compiles but then you get runtime errors.

A possible solution is:

```
if(NOT (${CMAKE_SYSTEM_NAME} MATCHES "Darwin"))
    CHECK_SYMBOL_EXISTS(aligned_alloc stdlib.h USE_ALIGNED_ALLOC)
endif()
```

 since posix_memalign seems to be there on macOS, and works well.

michaelld commented:
well drat! so we need a "runtime" check the symbol ... OSX can be so crappy sometimes!

jdemel commented:
Since AppVeyor passes, I assume we have a fix for Windows MSVC. At least it runs the generic kernels and thus, it must use `volk_malloc`.
But why would `CHECK_SYMBOL_EXISTS(aligned_alloc stdlib.h USE_ALIGNED_ALLOC)` basically return true if this is not the case? At the very least I'd expect a system to clearly communicate its capabilities.

@michaelld, @carlesfernandez  could you add a commit to this PR or publish it on GitHub that fixes the issue for MacOS? I don't have a MacOS box and thus, I can't test it.
I'm still

michaelld commented:
@jdemel yeah we'll figure out how to get OSX stuff working. Apple has a bad habit of using default compiler settings that don't barf on bad symbols, but either can't be linked or don't work in runtime.

carlesfernandez committed:
Fix for macOS

Although the aligned_alloc symbol is found in stdlib.h, it seems not implemented in macOS. This change makes macOS to choose the posix_memalign implementation.

Also fixes a mismatching parameter in endif()

carlesfernandez commented:
I pushed a change that seems to work well (runtime-tested on macOS 10.15 and 10.14) at https://github.com/jdemel/volk/pull/1

posix_memalign *should* work fine also in older macOS versions, but I have nowhere to test that.

michaelld commented:
:+1: from me to merge in @carlesfernandez's PR as noted. It will work with pretty much any usable OSX install.

jdemel committed:
Merge pull request #1 from carlesfernandez/aligned-alloc-fix-macos

Fix for macOS

jdemel commented:
Thanks for that fix @carlesfernandez I merged it. Let's see how our CI works and I assume this PR is ready to merge. fingers crossed.

michaelld reviewed (approved):
Works for me on OSX

michaelld commented:
Relevant CI passes. Maintainer approval. Merging.
